### PR TITLE
feat(do): lane partition, featured scoring, rings and clear-expired

### LINKED ISSUE #16: Do feature logic: lane partitioning, featured-Now scoring, clear-expired, auto-advance, rings math
Part of #1 (Kro Web parity epic).

**What.** `packages/app/src/features/do/**` (slice/Selectors/Shifters/Producers): the Do vista's in-memory regroup (`applyRegroup` equivalent) — Overdue (pending, due earlier today, due-asc) / Due Soon (within `nowThresholdHours`, plus ongoing) / Expired (due yesterday or earlier) / Next (later today >2h) / Anytime (no due, not ongoing) / Completed Today; the **featured Now lane**: canon scoring (overdue +100, ≤2h +50, ≤6h +25, any due +10, ongoing +30, has duration +5, sessionPoints>10 +3; zero excluded; tasks+habits only), odd count 3–9 by width, top scorer centered; badge count (overdue+expired) and "N left today"; **Clear Expired** (all overdue+expired, await every provider mutation, then one atomic refetch/install); **auto-advance** order (hero → now → overdue → expired → next → anytime; events/completed never); rings math (habits vs tasks due today incl. overdue, expired excluded, visibility-filter-independent); suggestions lane state (connect nudges, per-source dismiss); mark-complete mode state incl. backdated completion.

**Why.** The daily execution surface's business rules — the highest-traffic canon in the app.

**Scenario:** `react-uzf-v1` · Stack Matrix Scenario 5 (Web Only)
**UZF pattern & constraints:** `RC-1`–`RC-9`, `RC-36`; sections computed by Shifters over ONE vista (never separate queries).
Canon: `Kro/Application/Now/DoLanes.md`, `DoShifters`, `Kro/Models/Endeavors.swift` scoring, `docs/Features/DayProgressRings.md`, `docs/Features/Do.md`. Canon: `zheref/KroApple@2c1ee45` (re-fetch `origin/main` before building; note divergence in the PR).

**File lane (exclusive):** `packages/app/src/features/do/**` (logic files only; pages/components in #17).

**Blocked by:** #9, #10, #11 · **Blocks:** #17.

**Acceptance criteria**
1. Fixture days partition into lanes exactly per canon (edge cases: due-midnight boundaries, ongoing+overdue overlap).
2. Featured selection is odd-count, centered, deterministic for ties; zero-score exclusion holds.
3. Auto-advance picks the documented next card after each completion; rings ignore lens filters.

**Test expectations:** ≥3 per reducer arm/shifter/selector/producer; table-driven lane suite from the canon doc; clock-controlled tests for thresholds.
**Analytics/Distribution:** n/a.
**Out of scope:** rendering (#17), notifications delivery (#34).

### PR DESCRIPTION
**Pull request by zheref**
> 🟥 **Ichigo · Substitute** — the local plane, entire · *local, on your creds · I open PRs for **you** to merge (I never merge `main`, never review my own work)*
> acting as **Shinigami** (product code) · authored locally in a Claude Code session on behalf of @zheref — no CI run

Closes #16 · Part of #1

# What this changes for you

The Do tab's business rules now exist. Nothing is on screen yet — #17 owns the rendering — but the whole daily-execution engine is here and under test: give it a day's endeavors and an instant, and it tells you exactly what belongs in every lane, which card is the hero, how far round the rings have gone, what "N left today" says, what **Clear Expired** would close, and where focus lands after you tick something off.

Concretely, `packages/app/src/features/do/**` adds:

| Piece | What it decides |
|---|---|
| `DoRules.ts` | the lane partition — Overdue / Due Soon / Expired / Next / Anytime / Completed Today — the visibility terms, the auto-advance order and Clear Expired's target set |
| `DoFeaturedNow.ts` | the featured-Now score, the odd-count hero-centred arrangement, and the 3/5/7/9 responsive window |
| `DoRings.ts` | the two day-progress rings, and the rule that visibility filters never move them |
| `DoSuggestions.ts` | the connect nudges, per-source dismissal and the `do.showSuggestions` gate |
| `DoFeature.ts` | the slice: one `load` lifecycle field, the reconciled channels, the lanes, the preparation cursor, bulk mark-complete mode and the backdated-completion popover intent |
| `DoShifters.ts` / `DoSelectors.ts` / `DoProducer.ts` | the pure state transitions, the derived reads, and four thunks over the local store |
| `DoMocks.ts` | one fixture day carrying every lane and every boundary, plus ten canned `DoState` variants |

**Do is one vista.** `EndeavorsVistas.doTab` is fetched once, reconciled once (`#12`'s reconcile-before-grouping contract), and partitioned in memory by Shifters. There is no per-lane query.

**The clock is injected, everywhere.** No reducer, Shifter or Selector in this feature reads `Date.now()` or builds a `Date`: every event carries `now` in its payload and every thunk takes it as an argument. A test in `DoFeature.test.ts` greps the feature's own sources and fails if that ever stops being true. It is what makes a midnight-boundary case a plain unit test, and it is why `clockAnchor` is parked in state for the pure ring Selectors to read.

**One line in `store.ts`**, directly after the greeting entry, plus one flagged test-literal fix-up (below). No new dependencies, no `ThunkExtra` port: preferences, feature flags and endeavors all arrive through the `localStore` bundle that is already in the manifest.

**295 tests** across ten suites, all clock-controlled, none touching the network.

## How to verify

```bash
git fetch origin && git checkout ichigo/16-do-logic
make setup
make typecheck && make test && make lint && make build
```

All four are green on this branch (`@kro/app`: 599 tests across 25 files).

### Acceptance criterion 1 — fixture days partition into lanes exactly per canon, including due-midnight boundaries and the ongoing + overdue overlap

```bash
pnpm --filter @kro/app exec vitest run src/features/do/__tests__/DoRules.test.ts
```

`DoRules.test.ts` opens with the table-driven suite from `DoLanes.md`: 25 rows, one per fixture, each naming the lanes it must land in and why. The boundaries it pins, in order of how easy they are to get wrong:

- **Due at exactly `now`** → Due Soon, never Overdue (`isDueNow` is `due >= now`; `isOverdue` is `due < now`).
- **Due in exactly two hours** → Due Soon; **one second later** → Next. The window is inclusive on the Now side and strict on the Next side.
- **Midnight edge** — `moves a task from Overdue to Expired the moment the calendar day turns` walks the 30 seconds either side of midnight for a task due 23:59: nothing about the task changes, only which day `now` is on.
- **Due at 00:00 today** → Overdue (it is due *today*); **due at 00:00 tomorrow** → no lane at all.
- **Ongoing + overdue overlap** → Overdue alone. Never both; a separate case asserts no fixture in the day is ever matched by both predicates.
- **Ongoing, undated or due tomorrow** → Due Soon's ongoing tail, never Anytime.
- **Skipped** → no lane at all (`hasBeenCompleted` counts it; Completed Today does not).

`lane ordering` pins the sorts, including the one the doc gets wrong (see Notes).

### Acceptance criterion 2 — featured selection is odd-count, centred, deterministic for ties; zero-score exclusion holds

```bash
pnpm --filter @kro/app exec vitest run src/features/do/__tests__/DoFeaturedNow.test.ts
```

- **Scoring** — each weight is asserted at its own boundary: overdue `+100`, ≤2h `+50`, ≤6h `+25` (inclusive, with the moment past it dropping to `+10`), any due date `+10`, ongoing `+30`, duration `+5`, `sessionPoints > 10` `+3` (at exactly 10 it does *not* fire). Completed and reminder-kind endeavors score 0; a habit scores exactly as a task.
- **Centring** — three cards read `[2nd, 1st, 3rd]`; at seven the hero is at index 3; `centredFeaturedWindow` narrows by dropping the lowest-ranked flankers from *both* ends, so `selectDoFeaturedNowLane` at capacity 3 and at capacity 9 return windows with the same hero (`DoSelectors.test.ts`).
- **Odd count** — an even count of three or more loses its weakest card; a two-card day keeps both; the ceiling is nine.
- **Ties** — `breaks a score tie on the earlier due date`, and `is deterministic on a full tie` runs the same input twice and asserts an identical order.
- **Zero-score exclusion** — `excludes every zero-scoring endeavor`, and `is empty when nothing scores`.

### Acceptance criterion 3 — auto-advance picks the documented next card after each completion; rings ignore lens filters

```bash
pnpm --filter @kro/app exec vitest run src/features/do/__tests__/DoShifters.test.ts src/features/do/__tests__/DoRings.test.ts src/features/do/__tests__/DoSelectors.test.ts
```

- **Auto-advance order walk** — `walks the documented order after each completion until the day is empty` builds a three-card day, then loops *complete → advance* and asserts the full sequence of focused keys, ending in `null` (focus clears without jumping). `heroFirstSearchOrder` is pinned at 3 and at 9; `doAutoAdvanceLaneOrder` is asserted to be hero → now → overdue → expired → next → anytime, to exclude Completed Today, and to keep Expired beside Overdue. Bulk mark-complete mode always falls into the clearing branch, and the preference being off clears focus without jumping.
- **Rings ignore filters** — the two ring functions take the raw channels and *accept no lens argument at all*, which is the guarantee made structural. On top of that, `DoSelectors.test.ts` hides the habit kind, asserts the lanes lost every habit, and asserts the gold ring is byte-identical; the same is done for the Completed Today state.
- **Rings truth table** — `DoRings.test.ts` walks the five states from `DayProgressRings.md` (nothing expected / habits only / tasks only / both partly done / both complete), plus: no denominator ⇒ **absent**, not an empty track; overdue counted; expired excluded outright; undated tasks excluded; a habit never counted twice as a task; bulk mode and the `doActivityRings` kill switch hiding the readout.

### Clear Expired atomicity

```bash
pnpm --filter @kro/app exec vitest run src/features/do/__tests__/DoProducer.test.ts
```

Three tests carry it, against a stubbed in-memory store with an instrumented endeavor store:

- `awaits every mutation before the single refetch` — asserts on the call log that both writes precede the one read that follows them.
- `installs nothing at all when a mutation fails halfway — no partial day is observable` — subscribes to the store, fails the second write, and asserts that **every** state the reducer produced during the operation carried the same lanes. There is no instant at which half the day is cleared.
- `reports the refresh failure separately once the mutations have landed` — the two failure modes stay distinct, because the user's next move differs.

A fourth asserts a cleared row is `closed` with `completed` still `null`: clearing is an acknowledgement, not an achievement, and stamping it would let expired work fill a ring.

### Try it yourself

```bash
pnpm --filter @kro/app exec vitest run src/features/do --reporter=verbose
```

295 assertions, each named as a scenario. Nothing reaches the network; every suite goes through `makeStore(extra)` with a stubbed `localStore`.

### Canon

Re-fetched `zheref/KroApple` `origin/main` before starting: it is at **`2c1ee45`**, the same commit #1 pinned. **No canon divergence from the epic's pin.**

### Doc ↔ code divergences found in canon, and which one this follows

The epic's rule is that `docs/Features/*.md` binds and **code is the tie-breaker**. Three places where they disagree, all resolved toward the code, all worth a decision:

1. **Clear Expired's target set.** `DoLanes.md` § 6 says the action *"clears all `isOverdue` tasks (both today's Overdue and Expired)"*. The reducer arm says `isOverdue && !isDueToday` — the Expired lane **alone**. This port follows the code and leaves today's Overdue untouched. `doClearExpiredTargets` carries the reasoning and is the one line that changes if you rule for the doc.
2. **Completed Today's pre-sync clause.** `DoLanes.md` § 9 says rows *"marked complete in the current session but not yet synced (no `completed` date set yet) are also included"*. `EndeavorComputedState.completedToday` requires a completion date, and the Do shifter filters on nothing else — so on iOS a just-completed card leaves every lane until the refetch returns it stamped. This port makes the doc's promise true instead, by stamping the completion optimistically (deviation 4 below).
3. **Due Soon's sort.** `DoLanes.md` § 5 says the lane is *"sorted by due date ascending"*. The code sorts the due-window head and then **appends the ongoing tail unsorted**, so the lane is not globally due-ordered. Ported as the code has it, and pinned by a test.

### Boundary semantics the docs do not state, discovered in the Swift

Each is pinned by a test and documented at its definition:

- `isDueNow` is `due >= now` while `isOverdue` is `due < now`, so the instant a task falls due it is **Due Soon**, not Overdue.
- The Due Soon window is inclusive (`<= hours`) and Next is strictly `>`: exactly-two-hours-away is Due Soon.
- Overdue vs Expired splits on the **calendar day of the due date**, not a rolling 24h window — something due 23:59 yesterday is Expired at 00:01 today.
- `isDueNext` additionally requires `isDueToday`, so **a pending task due tomorrow appears in no lane at all** — not Next (not today), not Anytime (it has a due date). This is canon behaviour, not an omission here; it is the fixture `dueTomorrowMorning`.
- The ongoing tail admits an ongoing task *whatever* its due date, so an ongoing task due tomorrow does surface — in Due Soon.
- `hasBeenCompleted` also counts `skipped` / `reviewing` / `qa`, while Completed Today requires `status === closed`: a **skipped** endeavor therefore appears in **no** lane.
- The featured lane's 2h / 6h proximity bands are canon constants and are **not** wired to `do.nowThresholdHours`; the preference moves the lane boundary only.
- `selectFeaturedNowTasks` drops a card to keep the count odd only from three upward — a two-card day legitimately shows an even lane.

### Deliberate deviations from canon, and why

1. **The clock is a parameter.** Canon's `featuredNowScore` and `isOverdue` read `Date()` inside the property, so the same endeavor scores differently on consecutive reads and no test can pin a boundary. Every predicate here takes `now`, exactly as canon already does for `isDueNow(withinHours:now:)`. Weights, thresholds and ordering are unchanged.
2. **`resolvedKind`, not `kind`.** Canon guards every lane on the source-resolved kind. `packages/core/src/vistas` still reads plain `kind` with a doc note at each site saying adoption is a *vistas-lane* change — so this feature re-states the three computed-state guards locally against `resolvedKind` rather than editing files outside its lane. A daily reminder therefore participates here as a Habit, which is what `DayProgressRings.md` requires.
3. **Ties resolve stably.** Ranking is score-desc then due-asc; `Array.sort` is stable where Swift's `sorted` is not, so a full tie falls back to the reconciled pool's own first-appearance order instead of being arbitrary. Strictly more determinism, no change to the ordering canon specifies.
4. **The optimistic completion is stamped.** Canon sets only `status = .closed` and lets the host return the timestamp. Here the local store *is* the host and the persist writes exactly this instant, so stamping it keeps the optimistic and refetched states identical — and it is what makes both specs' promises true at the tap: the card enters Completed Today, and the arc sweeps forward. A completion backdated to an earlier day correctly does neither (tested).
5. **Visibility is stored as arrays, not an `EndeavorsLens`.** The lens carries five `Set`s and the store's `serializableCheck` admits `Date` and plain objects only. `DoState` holds the four toggles Do exposes as plain arrays and materialises the lens at the point of use (`doLensFor`), so the vista keeps owning `showArchived: true` and the exposed toggle set.
6. **Apple sources are absent, so Suggestions has one member.** The epic puts Apple Reminders and Apple Calendar permanently out of scope (no EventKit on web); Google Calendar is the flagship external host. `DoSuggestionSource` declares that one source, and dismissal is keyed on the **source** rather than on canon's copy-hash card id — so re-wording a nudge cannot resurrect one the user already turned down. Adding a source later is a member plus a branch.
7. **The Apple-Reminders relevance branch in the tasks ring is not ported.** It has no reachable input on web; the remaining `due`-is-today term is canon's own fallback for every other host.

### Scope

- **Events are installed, not laned.** The single fetch installs the calendar-event channel so nothing is discarded, but the all-day / timed grouping and the session-skip set from `DoLanes.md` § 2 are not among this issue's deliverables and are not built here.
- **No barrel export.** `packages/app/src/index.ts` is outside this issue's file lane, so the Do surface is not re-exported from `@kro/app` yet. #17 adds the exports it needs in its own lane.
- **No `use<F>` hook, no Page, no Fragment** — #17 owns the render layer.

### Flagged lane exceptions

- `packages/app/src/library/store.ts` — the one reducer entry, added **directly after** the greeting entry as the issue asked.
- `packages/app/src/features/greeting/__tests__/GreetingSelectors.test.ts` — a one-line test-literal fix-up. `RootState` now has a second slice, so that suite's hand-built root needs a `do:` field; it asserts nothing about it and says so in a comment. Same class as the #48 precedent, flagged here.

### Definition of Done — the items that are not fully satisfied

- **Coverage is argued, not measured.** `@kro/app` has no coverage provider installed (`@vitest/coverage-v8` is wired in `apps/web` only), and adding a dev dependency is outside this issue's lane and its "no new dependencies" constraint. Instead: every exported symbol in the feature is referenced by a test, every reducer arm / Shifter / Selector / Producer meets the ≥3 minimum, and the audit that found the four thinly-covered helpers (`isActionableDoKind`, `isComputedStateVisible`, `doAutoAdvanceLaneOrder`, `laneCards`) is reflected in the last three `describe` blocks of `DoRules.test.ts`. **Wiring coverage for `@kro/app` deserves its own change** — worth filing.
- **`UZF-26` visual evidence: exempt.** Logic-only session; no rendered UI surface changed.
- **`UZF-21` feature spec:** `docs/Features/Do.md` and `DayProgressRings.md` are canon in `zheref/KroApple`, not in this repo, and this change ports existing behaviour rather than adding any — so there is no spec here to update. Flag wrapping is `UZF-22`-clean: `doActivityRings` and `googleCalendar` are read through the flag registry and AND'd with their preferences where canon does.

---

Bankai-Agent: ichigo

### CODE REVIEW

File: packages/app/src/features/do/DoProducer.ts [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+ */
+export const fetchDoEndeavorsThunk = createAsyncThunk<
+  Result<DoDaySnapshot, DoException>,
+  { now: Date },
+  { extra: ThunkExtra }
+>('do/onEndeavorsFetchCompleted', async ({ now }, { extra }) => {
+  try {
+    const context = makeReconciliationContext({ now })
+    const stored = await readStoredEndeavors(extra.localStore)
+    return ok({ endeavors: reconcile(stored, context), now })
```
[the remark this answers is not in the document]
Comment (zheref) [reply packages/app/src/features/do/DoProducer.ts]: Fixed at ba44806 — both thunks now return the raw stored list and withEndeavorsInstalled owns the single reconcile pass (the #12 contract applied exactly once). All 603 @kro/app tests green with no assertion fallout, confirming install-side reconciliation was already the behavior under test.

File: packages/app/src/features/do/DoFeature.ts [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+    /**
+     * Lifecycle: the available width changed, so the featured lane may show a
+     * different odd count. One primitive field — the arrangement itself is
+     * untouched, which is what keeps the hero still across a resize.
+     */
+    onFeaturedCapacityChanged(
+      state,
+      action: PayloadAction<{ capacity: FeaturedNowCapacity }>,
+    ) {
+      state.featuredCapacity = action.payload.capacity
```
[the remark this answers is not in the document]
Comment (zheref) [reply packages/app/src/features/do/DoFeature.ts:220]: Fixed at ba44806 — same-value capacity assigns now return early, matching the documented no-op contract.

**Comment by zheref** [comment]
**Copilot round dispositioned** (2 threads, both fixed at `ba44806`: single reconcile pass at install; no-op capacity guard). Canon doc↔code divergences from this port are filed upstream as zheref/KroApple#497; the coverage-provider gap is kro-pwa#50.

**Fast-tracked** under the maintainer's express grant (mode → zheref/bankai-core#903): category (a) feature logic, no UI — merged by the orchestrating session.
